<?xml version="1.0"?>
<launch>


  <arg name="paused" default="false"/>
  <arg name="use_sim_time" default="true"/>
  <arg name="gui" default="true"/>
  <arg name="headless" default="false"/>
  <arg name="debug" default="false"/>

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find swarmbot)/config/model.world"/>
    <arg name="debug" value="$(arg debug)" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="paused" value="$(arg paused)"/>
    <arg name="use_sim_time" value="$(arg use_sim_time)"/>
    <arg name="headless" value="$(arg headless)"/>
  </include>

<!--#############################################################-->
  <node name="map_server" pkg="map_server" type="map_server" args="/home/hady/mapfiles/map.yaml">
    </node>
  
  <node pkg="tf" type="static_transform_publisher" name="world_map_static_broadcaster" args="0 0 0 0 0 0 1 world /map 100" />

<!--#############################################################-->

  <arg name="rviz_file1" default="$(find swarmbot)/rviz/rob1.rviz" />
  <arg name="rviz_file2" default="$(find swarmbot)/rviz/rob2.rviz" />
  <arg name="rviz_file3" default="$(find swarmbot)/rviz/rob3.rviz" />
  <node pkg="rviz" type="rviz" name="rviz_rob1" args="-d $(arg rviz_file1)" output="screen"/>
  <node pkg="rviz" type="rviz" name="rviz_rob2" args="-d $(arg rviz_file2)" output="screen"/>
  <node pkg="rviz" type="rviz" name="rviz_rob3" args="-d $(arg rviz_file3)" output="screen"/>
  




<!--#############################################################-->
<group ns="rob1">
<param name="tf_prefix" value="/rob1" />

<node pkg="tf" type="static_transform_publisher" name="map_odom_static_broadcaster" args="0 0 0 0 0 0 1 map /rob1/odom 100" />

<!-- node pkg="tf" type="static_transform_publisher" name="laser_base_rob1_static_broadcaster" args="-0.225 0 0 0 0 0 1 rob1/laser rob1/base_link 100" / -->
  
<param name="/rob1/robot_description" command="$(find xacro)/xacro --inorder '$(find swarmbot)/urdf/diff_wheeled_robot.xacro'" /> 
 
  <!-- start robot state publisher -->
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="rob1_state_publisher" output="screen" />

  <!-- push rob1_description to factory and spawn robot in gazebo -->
  <node name="spawn_rob1" pkg="gazebo_ros" type="spawn_model" args=" -x -5 -y -2 -urdf -model /rob1_model -param /rob1/robot_description" respawn="false" output="screen" />
 
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
  <param name="use_map_topic"             value="false"/>
  <!-- Publish scans from best pose at a max of 10 Hz -->
  <param name="odom_model_type" value="diff"/>
  <param name="odom_alpha5" value="0.1"/>
  <param name="transform_tolerance" value="0.2" />
  <param name="gui_publish_rate" value="10.0"/>
  <param name="laser_max_beams" value="30"/>
  <param name="min_particles" value="500"/>
  <param name="max_particles" value="5000"/>
  <param name="kld_err" value="0.05"/>
  <param name="kld_z" value="0.99"/>
  <param name="odom_alpha1" value="0.2"/>
  <param name="odom_alpha2" value="0.2"/>
  <!-- translation std dev, m -->
  <param name="odom_alpha3" value="0.8"/>
  <param name="odom_alpha4" value="0.2"/>
  <param name="laser_z_hit" value="0.5"/>
  <param name="laser_z_short" value="0.05"/>
  <param name="laser_z_max" value="0.05"/>
  <param name="laser_z_rand" value="0.5"/>
  <param name="laser_sigma_hit" value="0.2"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_model_type" value="likelihood_field"/>
  <!-- <param name="laser_model_type" value="beam"/> -->
  <param name="laser_likelihood_max_dist" value="2.0"/>
  <param name="update_min_d" value="0.2"/>
  <param name="update_min_a" value="0.5"/>
  <param name="odom_frame_id" value="odom"/>
  <param name="resample_interval" value="1"/>
  <param name="transform_tolerance" value="10.0"/>
  <param name="recovery_alpha_slow" value="0.0"/>
  <param name="recovery_alpha_fast" value="0.0"/>
  <param name="laser_frame" type="string" value="laser_link_rob1"/>
  <remap from="/cmd_vel" to="/rob1/cmd_vel"/>
  <remap from="/odom" to="/rob1/odom"/>
  <remap from="/particlecloud" to="/rob1/particlecloud"/>
  <remap from="/scan" to="/rob1/scan"/>
  <remap from="/amcl_pose" to="/rob1/amcl_pose"/>
  <remap from="/initialpose" to="/rob1/initialpose"/>
  <remap from="/rob1/static_map" to="/static_map"/>
  <!--<remap from="/particlecloud" to="/rob1_PF/particlecloud"/>
  <remap from="/estimated_pose" to="/rob1_PF/estimated_pose"/>
  <remap from="/MCL_estimated_pose" to="/rob1_PF/MCL_estimated_pose"/>-->
  </node>

   <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="controller_frequency" value="0.5"/> 
    <param name="controller_patiente" value="15.0"/>
    <rosparam file="$(find swarmbot)/param/mult/costmap_common_params_rob1.yaml" command="load" ns="global_costmap" /> 
    <rosparam file="$(find swarmbot)/param/mult/costmap_common_params_rob1.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find swarmbot)/param/mult/local_costmap_params_rob1.yaml" command="load" />
    <rosparam file="$(find swarmbot)/param/mult/global_costmap_params_rob1.yaml" command="load" /> 
    <rosparam file="$(find swarmbot)/param/mult/base_local_planner_params_rob1.yaml" command="load" />
    <remap from="move_base_simple/goal" to="rob1/move_base_simple/goal" />
    <remap from="/rob1/rob1/scan" to="/rob1/scan"/>

 </node>

  
</group>

<!--#############################################################-->
<group ns="rob2">
<param name="tf_prefix" value="/rob2" />


<node pkg="tf" type="static_transform_publisher" name="map_odom_static_broadcaster" args="0 0 0 0 0 0 1 map /rob2/odom 100" />

<!--node pkg="tf" type="static_transform_publisher" name="laser_base_rob2_static_broadcaster" args="-0.225 0 0 0 0 0 1 rob2/laser rob2/base_link 100" / -->
  

<param name="/rob2/robot_description" command="$(find xacro)/xacro --inorder '$(find swarmbot)/urdf/diff_wheeled_robot.xacro'" /> 
 
  <!-- start robot state publisher -->
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="rob2_state_publisher" output="screen" />

  <!-- push rob2_description to factory and spawn robot in gazebo -->
  <node name="spawn_rob2" pkg="gazebo_ros" type="spawn_model" args=" -x 2 -y 4 -Y -1.6 -unpause -urdf -model /rob2_model -param /rob2/robot_description" respawn="false" output="screen" />
 
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
  <param name="use_map_topic"             value="false"/>
  <!-- Publish scans from best pose at a max of 10 Hz -->
  <param name="odom_model_type" value="diff"/>
  <param name="odom_alpha5" value="0.1"/>
  <param name="transform_tolerance" value="0.2" />
  <param name="gui_publish_rate" value="10.0"/>
  <param name="laser_max_beams" value="30"/>
  <param name="min_particles" value="500"/>
  <param name="max_particles" value="5000"/>
  <param name="kld_err" value="0.05"/>
  <param name="kld_z" value="0.99"/>
  <param name="odom_alpha1" value="0.2"/>
  <param name="odom_alpha2" value="0.2"/>
  <!-- translation std dev, m -->
  <param name="odom_alpha3" value="0.8"/>
  <param name="odom_alpha4" value="0.2"/>
  <param name="laser_z_hit" value="0.5"/>
  <param name="laser_z_short" value="0.05"/>
  <param name="laser_z_max" value="0.05"/>
  <param name="laser_z_rand" value="0.5"/>
  <param name="laser_sigma_hit" value="0.2"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_model_type" value="likelihood_field"/>
  <!-- <param name="laser_model_type" value="beam"/> -->
  <param name="laser_likelihood_max_dist" value="2.0"/>
  <param name="update_min_d" value="0.2"/>
  <param name="update_min_a" value="0.5"/>
  <param name="odom_frame_id" value="odom"/>
  <param name="resample_interval" value="1"/>
  <param name="transform_tolerance" value="10.0"/>
  <param name="recovery_alpha_slow" value="0.0"/>
  <param name="recovery_alpha_fast" value="0.0"/>
  <param name="laser_frame" type="string" value="laser_link_rob2"/>
  <remap from="/cmd_vel" to="/rob2/cmd_vel"/>
  <remap from="/odom" to="/rob2/odom"/>
  <remap from="/particlecloud" to="/rob2/particlecloud"/>
  <remap from="/scan" to="/rob2/scan"/>
  <remap from="/amcl_pose" to="/rob2/amcl_pose"/>
  <remap from="/initialpose" to="/rob2/initialpose"/>
  <remap from="/rob2/static_map" to="/static_map"/>
  <!--<remap from="/particlecloud" to="/rob2_PF/particlecloud"/>
  <remap from="/estimated_pose" to="/rob2_PF/estimated_pose"/>
  <remap from="/MCL_estimated_pose" to="/rob2_PF/MCL_estimated_pose"/>-->
  </node>

   <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="controller_frequency" value="0.5"/> 
    <param name="controller_patiente" value="15.0"/>
    <rosparam file="$(find swarmbot)/param/mult/costmap_common_params_rob2.yaml" command="load" ns="global_costmap" /> 
    <rosparam file="$(find swarmbot)/param/mult/costmap_common_params_rob2.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find swarmbot)/param/mult/local_costmap_params_rob2.yaml" command="load" />
    <rosparam file="$(find swarmbot)/param/mult/global_costmap_params_rob2.yaml" command="load" /> 
    <rosparam file="$(find swarmbot)/param/mult/base_local_planner_params_rob2.yaml" command="load" />
    <remap from="move_base_simple/goal" to="rob2/move_base_simple/goal" />

 </node>

  
</group>

<!--#############################################################-->
<group ns="rob3">
<param name="tf_prefix" value="/rob3" />


<node pkg="tf" type="static_transform_publisher" name="map_odom_static_broadcaster" args="0 0 0 0 0 0 1 map /rob3/odom 100" />

<!-- node pkg="tf" type="static_transform_publisher" name="laser_base_rob3_static_broadcaster" args="-0.225 0 0 0 0 0 1 rob3/laser rob3/base_link 100" / -->
  

<param name="/rob3/robot_description" command="$(find xacro)/xacro --inorder '$(find swarmbot)/urdf/diff_wheeled_robot.xacro'" /> 
 
  <!-- start robot state publisher -->
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="rob3_state_publisher" output="screen" />

  <!-- push rob3_description to factory and spawn robot in gazebo -->
  <node name="spawn_rob3" pkg="gazebo_ros" type="spawn_model" args=" -x 5 -y -3 -Y 1.6  -unpause -urdf -model /rob3_model -param /rob3/robot_description" respawn="false" output="screen" />
 
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
  <param name="use_map_topic"             value="false"/>
  <!-- Publish scans from best pose at a max of 10 Hz -->
  <param name="odom_model_type" value="diff"/>
  <param name="odom_alpha5" value="0.1"/>
  <param name="transform_tolerance" value="0.2" />
  <param name="gui_publish_rate" value="10.0"/>
  <param name="laser_max_beams" value="30"/>
  <param name="min_particles" value="500"/>
  <param name="max_particles" value="5000"/>
  <param name="kld_err" value="0.05"/>
  <param name="kld_z" value="0.99"/>
  <param name="odom_alpha1" value="0.2"/>
  <param name="odom_alpha2" value="0.2"/>
  <!-- translation std dev, m -->
  <param name="odom_alpha3" value="0.8"/>
  <param name="odom_alpha4" value="0.2"/>
  <param name="laser_z_hit" value="0.5"/>
  <param name="laser_z_short" value="0.05"/>
  <param name="laser_z_max" value="0.05"/>
  <param name="laser_z_rand" value="0.5"/>
  <param name="laser_sigma_hit" value="0.2"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_model_type" value="likelihood_field"/>
  <!-- <param name="laser_model_type" value="beam"/> -->
  <param name="laser_likelihood_max_dist" value="2.0"/>
  <param name="update_min_d" value="0.2"/>
  <param name="update_min_a" value="0.5"/>
  <param name="odom_frame_id" value="odom"/>
  <param name="resample_interval" value="1"/>
  <param name="transform_tolerance" value="10.0"/>
  <param name="recovery_alpha_slow" value="0.0"/>
  <param name="recovery_alpha_fast" value="0.0"/>
  <param name="laser_frame" type="string" value="laser_link_rob3"/>
  <remap from="/cmd_vel" to="/rob3/cmd_vel"/>
  <remap from="/odom" to="/rob3/odom"/>
  <remap from="/particlecloud" to="/rob3/particlecloud"/>
  <remap from="/scan" to="/rob3/scan"/>
  <remap from="/amcl_pose" to="/rob3/amcl_pose"/>
  <remap from="/initialpose" to="/rob3/initialpose"/>
  <remap from="/rob3/static_map" to="/static_map"/>
  <!--<remap from="/particlecloud" to="/rob3_PF/particlecloud"/>
  <remap from="/estimated_pose" to="/rob3_PF/estimated_pose"/>
  <remap from="/MCL_estimated_pose" to="/rob3_PF/MCL_estimated_pose"/>-->
  </node>

   <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="controller_frequency" value="0.5"/> 
    <param name="controller_patiente" value="15.0"/>
    <rosparam file="$(find swarmbot)/param/mult/costmap_common_params_rob3.yaml" command="load" ns="global_costmap" /> 
    <rosparam file="$(find swarmbot)/param/mult/costmap_common_params_rob3.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find swarmbot)/param/mult/local_costmap_params_rob3.yaml" command="load" />
    <rosparam file="$(find swarmbot)/param/mult/global_costmap_params_rob3.yaml" command="load" /> 
    <rosparam file="$(find swarmbot)/param/mult/base_local_planner_params_rob3.yaml" command="load" />
    <remap from="move_base_simple/goal" to="rob3/move_base_simple/goal" />

 </node>

  
</group>

</launch>
